### PR TITLE
Fix repository search function

### LINKED ISSUE #2399: Dashboard shows all organization repositories, not respecting access rights
- Gitea version (or commit ref): f61a1d21
- Git version:
- Operating system:
- Database (use `[x]`):
  - [X] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [X] Not relevant
- Log gist:

## Description

See title.

### PR DESCRIPTION
**Pull request by Morlinest**
Rewrite of `models.SearchRepositoryByName`:

1. Adds respect of organizations membership visibility
2. Removes redudant `where` conditions in search (that was causing showing private repositories not available to `OwnerID`)
3. Includes "collaborative" repositories only if `opts.Collaborate` is set (`true`)
4. Removes need for `Searcher` as this option is confusing (you can't know, if found repository is related to `OwnerID` or `Searcher`) and not used if `OwnerID` is not set
5. Fixes `FIXME` parts mentioned in unit and integration tests (`integrations/api_repo_test.go` and `models/repo_list_test.go`)
6. Adds unit tests to test organizations membership visibility
7. Adds integration tests

Important files are `routers/api/v1/repo/repo.go` and `models/repo_list.go`, other files are for tests.

Fixes #2399, #2593.

Main/final part of #2371.

### CODE REVIEW

**Comment by Morlinest** [comment]
One thing I found when testing is that team with read access to public repository is not listed because members are not added in `access` table. I think we should disable this option (team with read access) when repository is public.

File: models/repo_list.go [hunk 1]
```diff
@@ -136,48 +136,44 @@ const (
...
 			}
+
+			cond = cond.And(accessCond)
 		}
-		cond = cond.And(searcherReposCond)
 	}
 
-	if !opts.Private {
-		cond = cond.And(builder.Eq{"is_private": false})
+	if opts.Keyword != "" {
```
Comment (ethantkoenig) [inline models/repo_list.go]: This is a duplicate of lines 183-186

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
It seems that `opts.Searcher` is no longer used? If this is intentional, then we should remove it from the struct.

File: models/repo_list.go (hunk 1 above)
Comment (Morlinest) [reply models/repo_list.go]: Fixed (rebase doesn't work always as expected :D)

**Comment by Morlinest** [comment]
@ethantkoenig I wanted to completely remove `Searcher` from code in another PR.

**Comment by ethantkoenig** [comment]
@Morlinest why?

**Comment by Morlinest** [comment]
Was not point of this PR, but as this is last function that was using it and it's only few lines to remove I can do it here.

**Comment by Morlinest** [comment]
@ethantkoenig `Searcher` removed from code completely

**Comment by Morlinest** [comment]
@ethantkoenig Is it OK now? Can you aproove changes?

**Comment by lafriks** [comment]
LGTM

**Comment by lafriks** [comment]
@ethantkoenig your approval needed

**Comment by ethantkoenig** [comment]
@Morlinest @lafriks I will take a look tonight

File: routers/api/v1/repo/repo.go [hunk 2]
```diff
@@ -34,31 +34,30 @@ func Search(ctx *context.APIContext) {
...
+	if opts.OwnerID > 0 {
+		var repoOwner *models.User
+		if ctx.User != nil && ctx.User.ID == opts.OwnerID {
+			repoOwner = ctx.User
 		} else {
-			u, err := models.GetUserByID(opts.OwnerID)
+			var err error
+			repoOwner, err = models.GetUserByID(opts.OwnerID)
 			if err != nil {
 				ctx.JSON(500, api.SearchError{
```
Comment (ethantkoenig) [inline routers/api/v1/repo/repo.go:46]: Should it be 404 if user doesn't exist?

File: models/fixtures/team.yml [hunk 3]
```diff
@@ -44,5 +44,6 @@
   name: Owners
   authorize: 4 # owner
   num_repos: 2
-  num_members: 1
+  num_members: 2
   unit_types: '[1,2,3,4,5,6,7]'
+
```
Comment (ethantkoenig) [inline models/fixtures/team.yml]: nit: remove trailing space

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/repo.go (hunk 2 above)
Comment (lafriks) [reply routers/api/v1/repo/repo.go:46]: @ethantkoenig don't thinks so

Comment (ethantkoenig) [reply routers/api/v1/repo/repo.go:46]: @lafriks IMO, it doesn't make sense to return a 500 if the client provides a bad user ID

Comment (lafriks) [reply routers/api/v1/repo/repo.go:46]: @ethantkoenig yes but not 404. It should be some other error or just empty repo list

Comment (daviian) [reply routers/api/v1/repo/repo.go:46]: @lafriks @ethantkoenig Sounds like a usual 400 Bad Request for me.

Comment (Morlinest) [reply routers/api/v1/repo/repo.go:46]: I didn't change it as it is current behaviour. I'll change it to "bad request".

Comment (Morlinest) [reply routers/api/v1/repo/repo.go:46]: @ethantkoenig @daviian I see problem, this response is defined in api (swagger):
```
	//     Responses:
	//       200: SearchResults
	//       500: SearchError
```

Comment (Morlinest) [reply routers/api/v1/repo/repo.go:46]: I would keep it as it is now and make changes in v2 api. I can fix only leaking internal information from response. Do you agree?

Comment (lafriks) [reply routers/api/v1/repo/repo.go:46]: @Morlinest @ethantkoenig as this is documented so let's not introduce breaking changes in API

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

File: routers/api/v1/repo/repo.go (hunk 2 above)
Comment (ethantkoenig) [reply routers/api/v1/repo/repo.go:46]: IMO, calling this a breaking change seems like a stretch, but I don't want to block this PR so I'll accept

**Pull request merged**